primitives/src/crh/poseidon/mod.rs

```diff
@@ -1,427 +1,213 @@
-        let m_21 = P::MDS_CST[3];
-        let m_22 = P::MDS_CST[4];
-        let m_23 = P::MDS_CST[5];
-
-        let elem_3 = state[0].mul(&m_21);
-        let elem_4 = state[1].mul(&m_22);
-        let elem_5 = state[2].mul(&m_23);
-
-        new_state[1] = elem_3;
-        new_state[1] += &elem_4;
-        new_state[1] += &elem_5;
-
-        // scalar multiplication for the position 2 of the state vector
-        let m_31 = P::MDS_CST[6];
-        let m_32 = P::MDS_CST[7];
-        let m_33 = P::MDS_CST[8];
-
-        let elem_6 = state[0].mul(&m_31);
-        let elem_7 = state[1].mul(&m_32);
-        let elem_8 = state[2].mul(&m_33);
-
-        new_state[2] = elem_6;
-        new_state[2] += &elem_7;
-        new_state[2] += &elem_8;
-
-        // copy the result to the state vector
-        state[0] = new_state[0];
-        state[1] = new_state[1];
-        state[2] = new_state[2];
-
-    }
-
-    // Function that does the mix matrix with fast algorithm
-    // It uses a partial Montgomery multiplication defined as PM(x, t) = x * t * 2^-64 mod M
-    // t is a 64-bit matrix constant. In the algorithm, the constants are represented in
-    // partial Montgomery representation, i.e. t * 2^64 mod M
-    fn matrix_mix_short (state: & mut Vec<F>) {
-
-        // the new state where the result will be stored initialized to zero elements
-        let mut new_state = vec![F::zero(); P::T];
-
-        let mut m_11 = P::MDS_CST_SHORT[0];
-        let mut m_12 = P::MDS_CST_SHORT[1];
-        let mut m_13 = P::MDS_CST_SHORT[2];
-
-        m_11.mul_short_assign(state[0]);
-        m_12.mul_short_assign(state[1]);
-        m_13.mul_short_assign(state[2]);
-
-        new_state[0] = m_11;
-        new_state[0] += &m_12;
-        new_state[0] += &m_13;
-
-        // scalar multiplication for position 1 of the state vector
-        let mut m_21 = P::MDS_CST_SHORT[3];
-        let mut m_22 = P::MDS_CST_SHORT[4];
-        let mut m_23 = P::MDS_CST_SHORT[5];
-
-        m_21.mul_short_assign(state[0]);
-        m_22.mul_short_assign(state[1]);
-        m_23.mul_short_assign(state[2]);
-
-        new_state[1] = m_21;
-        new_state[1] += &m_22;
-        new_state[1] += &m_23;
-
-        // scalar multiplication for the position 2 of the state vector
-        let mut m_31 = P::MDS_CST_SHORT[6];
-        let mut m_32 = P::MDS_CST_SHORT[7];
-        let mut m_33 = P::MDS_CST_SHORT[8];
-
-        m_31.mul_short_assign(state[0]);
-        m_32.mul_short_assign(state[1]);
-        m_33.mul_short_assign(state[2]);
-
-        new_state[2] = m_31;
-        new_state[2] += &m_32;
-        new_state[2] += &m_33;
-
-        // copy the result to the state vector
-        state[0] = new_state[0];
-        state[1] = new_state[1];
-        state[2] = new_state[2];
     }
 
-    fn poseidon_perm (state: & mut Vec<F>) {
-
-        let use_fast = true;
+    /// Inversion S-Box only !
+    pub(crate) fn poseidon_perm (state: &mut Vec<F>) {
```

Comment:
This function duplicates the one for batch evaluations of the x^-1 Permutation.
Besides that it is not as nicely structured into full_round and partial_round, why do we do that here once again?